### ISSUE TITLE
Rethinking the grid system research

### ISSUE THREAD

**Issue by msecret** [uswds/uswds#2137]
After having talked to some users about the WDS grid, I've found the following problems consistently mentioned:

### Documented offset classes

While there are offset classes for the grid system, they aren't documented on the main WDS site, leading to confusion as to whether they even exist.

**Why it's a problem**: Users expect offset classes to be there due to them being prevalent in other grid systems and frameworks.

### More then max columns requires additional `usa-grid` containers

**Why it's a problem**: 

- In dynamic code systems, content could be of variable amounts, so it's much easier to code for a grid system that will automatically adapt 5 vs 3 one-third columns.
- Even when it's not a dynamic code system, the users shouldn't have to think about when they need to add a grid container

### Can't nest grid columns

**Why it's a problem**:
- Certain layouts are much easier to do with nesting. A 1/4 column next to a 3/4 column, where the first element in the 3/4 column expands the full width, and it then splits up into 3 1/3 columns to fill the rest of the 3/4 column. Creating this layout with nested grids is as simple as:

```
.row-4          .row-4
content         large element
                .row-3  .row-3  .row-3
```

With the WDS grid, where columns can't be nested, it would mean:

```
.width-one-fourth     .width-three-fourths
content               large element
.usa-grid
.offset-one-forth       .width-one-fourth       .width-one-fourth       .width-one-fourth
```

### Can't target specific breakpoints

**Why it's a problem**: In general, elements is supposed to break and flow based on it's content, not predefined breakpoints. The Standards offers no way to do different things with the grid at different breakpoints, leading to inflexibility with the content.

### The naming convention is confusing

User's have stated:
- Unsure of when an "s" is needed on the end of width classes, `usa-two-halves`, `usa-one-half`.
- Hard to remember the name of things, such as `usa-grid-full`.

**Why it's a problem**: Difficulty in remember names of classes means it's a harder experience for developers to work with the grid, requiring looking up documentation to complete code.

**Comment by msecret** [comment]
Criteria:
A. Documented offset classes
B. More then max columns requires additional usa-grid containers
C. Can't nest grid columns
D. The naming convention is confusing

[BASSCSS](http://basscss.com/#basscss-grid)
A: **fail**, there are not offset classes in the grid.
B: **pass**, can add as many columns and things will flow correctly to next lines
C: **pass**, can nest
D: **unsure**, the system is more widely used but the naming convention is somewhat atomic, which isn't necessarily in the WDS style.

**Comment by msecret** [comment]
[Bootstrap](https://v4-alpha.getbootstrap.com/layout/grid/)
A: **pass**, there are offset classes in the grid.
B: **fail**, I think everything has to be nested in a `row` class
C: **pass**, can nest
D: **pass**, a well known naming convention

**Comment by msecret** [comment]
[reflex](http://leejordan.github.io/reflex/docs/)
A: **fail** there are no offset classes.
B: **pass** can use an `auto` column class to automatically flow as many columns as needed. Not the default though
C: **pass** says it supports nested grids but doesn't have any examples
D: **maybe** not as used and some things are common in other grid systems, like the `auto` class. Is also relatively complex compared to other grid systems.

**Comment by joshbruce** [comment]
If I saw the #2071 correctly, would flex-box also be a possibility now? 

Along with [Element UI](https://semantic-ui.com) we might also want to check out [Turret CSS](http://turretcss.com). Neat also has a flex-box switch that can be flipped, if memory serves. [Foundation](https://foundation.zurb.com/sites/docs/xy-grid.html) is kicking around and now has something they're calling XY Grid (dropped June 27th); release notes say it's based on [w3c css-grid](https://www.w3.org/TR/css-grid-1/).

Turret is written in Less, but I did find a converter to Sass somewhere. My big takeaway from some of these libraries is the emphasis on mixins. That, of course, has its own plusses and minuses.

2017 is starting to feel like 2005. Only instead of focusing on HTML everyone seems focused on CSS. :)

**Comment by msecret** [comment]
@joshbruce yea flexbox is a possibility, but I think we'd have to do more research on how well IE11 supports flexbox, because it's marked as only partial support and I've heard there are problems and bugs. If we could find a way to get flexbox working well IE11, it would definitely work. We could even potentially look at CSS grids, although I think they have even more issues with IE11.

Thanks for the links! I'm also hesitant on the mixin structure, as that's what our current grid is and there's been some problems with it.

**Comment by buckaroogeek** [comment]
I would like to see a css grid and flexbox variant of the standards.

**Comment by joshbruce** [comment]
@msecret: Found this recently - https://github.com/rachelandrew/gridbugs

I've abandoned IE11 support for my own work. The gov't program I'm on is also looking to use CSS Grid instead of float and flexbox - I think it's only advanced features that seem to be sketchy on IE11 - and there is a polyfill or two, I think.

Haven't had any negative feedback regarding the sites I've created using CSS Grid; so, there might be visual differences, but not enough to make people freak out.

**Comment by thisisdano** [comment]
@buckaroogeek me too

**Comment by maya** [comment]
Some feedback from the USWDS Slack around CSS Grid:

>since it's (supposed to be supported in all browsers), it would like eliminate any if* statements. Also, I am just learning about it, but it's nice that empty divs that hold their width & height can be used and we don't have to write separately for mobile.  I used to have issues with grids within grids ... but this new (7 month old) thing may be an interesting solution

**Comment by maya** [comment]
Implemented in #2422. Will be available in the next 2.0 release. Thanks for your thoughts, everyone!
